Empty condition in configuration crashes beats

When a `condition` from the configuration has no sub-fields,  `NewCondition` will return a `nil` condition. However, this `nil` condition will result in a panic when any of its methods is invoked.

- Version: 6.2.4
- Operating System: N/A
- Discuss Forum URL: https://discuss.elastic.co/t/filebeats-autodiscover-segmentation-violation/133342
- Steps to Reproduce: 

Use this broken config from the discuss thread above (note that the `equals` clause is wrongly indented at the same level as `condition`):
```
filebeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
          equals:
            docker.container.image: nginx
          config:
            - module: nginx
              access:
                prospector:
                  type: docker
                  container.ids:
                    - "${data.docker.container.id}"
              error:
                prospector:
                  type: docker
                  container.ids:
                    - "${data.docker.container.id}"
```

will result in a panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xebfea6]
goroutine 58 [running]:
github.com/elastic/beats/libbeat/processors.(*Condition).Check(0x0, 0x1f36d60, 0xc42007d5c0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/condition.go:205 +0x26
github.com/elastic/beats/libbeat/autodiscover/template.(*Mapper).GetConfig(0xc42049c380, 0xc42007d5c0, 0xc4204bc528, 0xac369f, 0x10)
	/go/src/github.com/elastic/beats/libbeat/autodiscover/template/config.go:62 +0x63e
github.com/elastic/beats/libbeat/autodiscover/providers/docker.(*Provider).publish(0xc4205340c0, 0xc42007d5c0)
	/go/src/github.com/elastic/beats/libbeat/autodiscover/providers/docker/docker.go:137 +0x3d
github.com/elastic/beats/libbeat/autodiscover/providers/docker.(*Provider).emitContainer(0xc4205340c0, 0xc42015a240, 0x16c6c00, 0x5)
	/go/src/github.com/elastic/beats/libbeat/autodiscover/providers/docker/docker.go:110 +0x78a
github.com/elastic/beats/libbeat/autodiscover/providers/docker.(*Provider).Start.func1(0xc4205340c0)
	/go/src/github.com/elastic/beats/libbeat/autodiscover/providers/docker/docker.go:74 +0x1e9
created by github.com/elastic/beats/libbeat/autodiscover/providers/docker.(*Provider).Start
	/go/src/github.com/elastic/beats/libbeat/autodiscover/providers/docker/docker.go:65 +0x3f
```

Apart from NewCondition always returning an usable condition, we should try to warn the user or err on this configuration mistake

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -64,6 +64,7 @@ https://github.com/elastic/beats/compare/v6.2.3...master[Check the HEAD diff]
 - Preserve the event when source matching fails in `add_docker_metadata`. {pull}7133[7133]
 - Negotiate Docker API version from our client instead of using a hardcoded one. {pull}7165[7165]
 - Fix delays on autodiscovery events handling caused by blocking runner stops. {pull}7170[7170]
+- Error out on invalid Autodiscover template conditions settings. {pull}7200[7200]
 
 *Auditbeat*
 

--- a/libbeat/autodiscover/appenders/config/config.go
+++ b/libbeat/autodiscover/appenders/config/config.go
@@ -45,11 +45,14 @@ func NewConfigAppender(cfg *common.Config) (autodiscover.Appender, error) {
 	var configMaps []configMap
 	for _, conf := range confs {
 		// Unpack the condition
-		cond, err := processors.NewCondition(conf.ConditionConfig)
+		var cond *processors.Condition
 
-		if err != nil {
-			logp.Warn("config", "unable to create condition due to error: %v", err)
-			continue
+		if conf.ConditionConfig != nil {
+			cond, err = processors.NewCondition(conf.ConditionConfig)
+			if err != nil {
+				logp.Warn("config", "unable to create condition due to error: %v", err)
+				continue
+			}
 		}
 		cm := configMap{condition: cond}
 

--- a/libbeat/autodiscover/autodiscover.go
+++ b/libbeat/autodiscover/autodiscover.go
@@ -3,6 +3,8 @@ package autodiscover
 import (
 	"time"
 
+	"github.com/pkg/errors"
+
 	"github.com/elastic/beats/libbeat/autodiscover/meta"
 	"github.com/elastic/beats/libbeat/beat"
 	"github.com/elastic/beats/libbeat/cfgfile"
@@ -60,7 +62,7 @@ func NewAutodiscover(name string, pipeline beat.Pipeline, adapter Adapter, confi
 	for _, providerCfg := range config.Providers {
 		provider, err := Registry.BuildProvider(bus, providerCfg)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "error in autodiscover provider settings")
 		}
 		logp.Debug(debugK, "Configured autodiscover provider: %s", provider)
 		providers = append(providers, provider)

--- a/libbeat/autodiscover/template/config_test.go
+++ b/libbeat/autodiscover/template/config_test.go
@@ -65,3 +65,21 @@ func TestConfigsMapping(t *testing.T) {
 		assert.Equal(t, test.expected, res)
 	}
 }
+
+func TestNilConditionConfig(t *testing.T) {
+	var mappings MapperSettings
+	data := `
+- config:
+    - type: config1`
+	config, err := common.NewConfigWithYAML([]byte(data), "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := config.Unpack(&mappings); err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = NewConfigMapper(mappings)
+	assert.Error(t, err)
+}

--- a/libbeat/processors/condition.go
+++ b/libbeat/processors/condition.go
@@ -68,7 +68,7 @@ func NewCondition(config *ConditionConfig) (*Condition, error) {
 
 	if config == nil {
 		// empty condition
-		return nil, nil
+		return nil, errors.New("missing condition config")
 	}
 
 	var err error

--- a/libbeat/processors/condition_test.go
+++ b/libbeat/processors/condition_test.go
@@ -57,6 +57,12 @@ func TestCreateConditions(t *testing.T) {
 	}
 }
 
+func TestCreateNilCondition(t *testing.T) {
+	cond, err := NewCondition(nil)
+	assert.Nil(t, cond)
+	assert.Error(t, err)
+}
+
 func GetConditions(t *testing.T, configs []ConditionConfig) []Condition {
 	conds := []Condition{}
 

--- a/metricbeat/autodiscover/appender/kubernetes/token/token.go
+++ b/metricbeat/autodiscover/appender/kubernetes/token/token.go
@@ -32,10 +32,13 @@ func NewTokenAppender(cfg *common.Config) (autodiscover.Appender, error) {
 		return nil, fmt.Errorf("unable to unpack config due to error: %v", err)
 	}
 
-	// Attempt to create a condition. If fails then report error
-	cond, err := processors.NewCondition(conf.ConditionConfig)
-	if err != nil {
-		return nil, fmt.Errorf("unable to create condition due to error: %v", err)
+	var cond *processors.Condition
+	if conf.ConditionConfig != nil {
+		// Attempt to create a condition. If fails then report error
+		cond, err = processors.NewCondition(conf.ConditionConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create condition due to error: %v", err)
+		}
 	}
 	appender := tokenAppender{
 		TokenPath: conf.TokenPath,

